src/plugins/intel_npu/src/al/include/npu.hpp

```diff
@@ -52,22 +53,12 @@ class IEngineBackend : public std::enable_shared_from_this<IEngineBackend> {
 
 //------------------------------------------------------------------------------
 
-class IExecutor {
-public:
-    virtual ~IExecutor() = default;
-
-    virtual void setWorkloadType(const ov::WorkloadType workloadType) const = 0;
-};
-
-//------------------------------------------------------------------------------
-
 class IDevice : public std::enable_shared_from_this<IDevice> {
 public:
     using Uuid = ov::device::UUID;
 
-    virtual std::shared_ptr<IExecutor> createExecutor(
-        const std::shared_ptr<const NetworkDescription>& networkDescription,
-        const Config& config) = 0;
+    virtual std::shared_ptr<IExecutor> createExecutor(const NetworkDescription& networkDescription,
```

Comment:
Replacing shared pointer (here and in other places) with raw reference is potentially dangerous operation.
Objects (e.g., executor) created and kept by shared pointer can outlive the ones used for their creation. 
Therefore, access to dead memory area is possible.

Thorough testing of these changes is required. CI checks may be not enough, there are applications which can deviate in structure and workflow from the benchmark_app and therefore be affected without immediate report from CI checks.
Benefit from these replacements is subtle, no visible cleanup is achieved.

Reply:
Added a test case that destroy each object after using it.

Reply:
Having the article seen on the internet does not make it absolute truth.
This particular article is mistaken in its statements about `void mayShare(const std::shared_ptr<Widget>& shaWid)`

The tests are of course good to have, but mentioned testing of the applications is still required. All the applications are potentially affected.
Lifecycle issues are still possible, I explained why in the start message of the topic.

Should someone observes the cleanup effect, I'd be glad too see these verbal explanations.

Reply:
I do observe the cleanup effect. Here is my explanation.

Now, looking at the current method signature, I see that `networkDescription` is read-only, and passing it to `createExecutor` won’t create any new owners of the object.

The previous method signature raises two questions:

1. Is `networkDescription` meant to be a shared object among multiple owners? Does `createExecutor` care about this?
2. Does `createExecutor` introduce a new owner of the object, or does it leave ownership unchanged?"

For me, the API became clearer at this place.

Reply:
Having the changes in place, are these questions removed, or some answers to them are given?

Reply:
Sure, let me comment if it is not clear.

> Is networkDescription meant to be a shared object among multiple owners? Does createExecutor care about this?

With having a const reference passed as an argument, this question is not needed to be asked. So yes, this question can be removed.

> Does createExecutor introduce a new owner of the object, or does it leave ownership unchanged?"

With having a const reference passed as an argument, it is obvious that `createExecutor` cannot change ownership of the object.

I see that we have different points of view on this issue. So I suggest adding a third side to this thread, so they choose what API is preferable in this case. Do you agree @alpkn ?

Reply:
Thanks for keeping on with the clarity.

>With having a const reference passed as an argument, this question is not needed to be asked. So yes, this question can be removed.

No matter whether `networkDescription` is passed as a shared pointer, raw reference or a pointer, it still can be shared with someone else by the receiver of such a reference. So in this sense the question is meaningless.

>With having a const reference passed as an argument, it is obvious that `createExecutor` cannot change ownership of the object.

Being a bit more precise, `createExecutor` (receiver) *can not increase* the count of shared pointer owners when not having the shared pointer itself. So, I read the intention of the change as following: a receiver must be able to use the `networkDescription` object, but not be present in the list of its shared pointer owners. Is that correct reading?

I invite anyone who is interested in the topic and feel competent to participate in discussion.

Reply:
> Is that correct reading?

Yes, it is.

Reply:
>a receiver must be able *to use* the networkDescription object, but not be present in the list of its shared pointer owners

With the changes, is there a guarantee that all the uses of the particular resource *happen before* all the ownerships of it are released? `networkDescription` is an exemplar resource here.